### ISSUE TITLE
Parsing erros for fields at the end of the record

### ISSUE THREAD

**Issue by fs-granma** [microsoft/krabsetw#87]
Hi,

I've ran into some problems while consuming events from the Microsoft-Windows-WinINet provider ({43d1a55c-76d6-4f7e-995c-64c711e5cafe}) on Windows 10.

Sometimes the last properties in the record cannot be parsed.

Sometimes the problematic properties (e.g. Event id: 1008 Prop: StatusLine, Event id: 506 Prop: URL) seem to be zero-length, non-null-terminated strings at the end of the record (i.e. there is a record with zero bytes of data). These hit the assertion 'invariant: if we've exhausted our buffer, then we must've exhausted the properties as well'. It seems that the end of the buffer might not strictly mean the end of properties.

There are also integer types that either throw an exception ('Property length past end of property buffer') or hit the assertion.

I have not observed similar issues with other providers so far.

Please find attached a list of event ids and property names from the Microsoft-Windows-WinINet provider that I have observed causing problems. Note that these don't always cause problems, but some properties (StatusLine, URL) seem to have a high probability to hit the assertion.

--Markus
[Microsoft-Windows-WinINet_parsing_issues.txt](https://github.com/microsoft/krabsetw/files/3859294/Microsoft-Windows-WinINet_parsing_issues.txt)

**Comment by fs-granma** [comment]
Hello,

Any ideas how to proceed with this?

I'm thinking that if it's the provider (Microsoft-Windows-WinINet in this case) that produces invalid data then the library should probably handle it somehow gracefully rather than asserting.

It also seems that the assertion 'invariant: if we've exhausted our buffer, then we must've exhausted the properties as well' should probably be somehow relaxed since it seems like a legitimate case of having a zero-length string at the end of the buffer.

Having integer types past the end of the buffer seems a bit suspicious though.

Thanks,
Markus

**Comment by jrave** [comment]
Hello,

we are still looking for a solution to this and hope that you can provide some guidance for this. At the end there is a small sample program that reproduces the behavior. The easiest for us has been to use Edge on Win 10 to trigger the issue.

What it seems like to us that some properties come in pairs <Name> and <Name>Length. If <Name>Length == 0 then <Name> should not even be attempted to parse. Question is now whether there can be done something in krabsetw to help with this or whether this should be purely handled on the application side.

Best Regards,
Johannes

```cpp
#include <iostream>
#include "..\..\krabs\krabs.hpp"
#include "examples.h"

void user_trace_wininet::start()
{
    // user_trace instances should be used for any non-kernel traces that are defined
    // by components or programs in Windows.
    krabs::user_trace trace;

    krabs::provider<> provider(krabs::guid(L"{43d1a55c-76d6-4f7e-995c-64c711e5cafe}"));

    // user_trace providers typically have any and all flags, whose meanings are
    // unique to the specific providers that are being invoked. To understand these
    // flags, you'll need to look to the ETW event producer.
    //provider.any(0xf0010000000003ff);

    // providers should be wired up with functions (or functors) that are called when
    // events from that provider are fired.
    provider.add_on_event_callback([](const EVENT_RECORD& record, const krabs::trace_context& trace_context) {

        // Once an event is received, if we want krabs to help us analyze it, we need
        // to snap in a schema to ask it for information.
        krabs::schema schema(record, trace_context.schema_locator);

        const auto event_id = schema.event_id();
        if (event_id == 1008) {
            krabs::parser parser(schema);

            // We then have the ability to ask a few questions of the event.
            std::wcout << L"Event " << schema.event_id();
            std::wcout << L"(" << schema.event_name() << L") received." << std::endl;

            for (const auto& prop : parser.properties())
            {
                switch (prop.type())
                {
                case TDH_INTYPE_UINT32:
                {
                    auto propValue = parser.parse<uint32_t>(prop.name());
                    std::wcout << L"\t" << prop.name() << ": " << propValue << std::endl;
                }
                break;
                case TDH_INTYPE_ANSISTRING:
                {
                    auto propValue = parser.parse<std::string>(prop.name());
                    printf("\t%ws: %hs\n", prop.name().c_str(), propValue.c_str());
                }
                break;
                default:
                    std::wcout << L"\tUnsupported: " << prop.name() << L" " << prop.type() << std::endl;
                }
            }


        }

        });

    // the user_trace needs to know about the provider that we've set up.
    trace.enable(provider);

    // begin listening for events. This call blocks, so if you want to do other things
    // while this runs, you'll need to call this on another thread.
    trace.start();
}
```

**Comment by jrave** [comment]
Created PR #105 with some fixes that should be able to help with this. Please have a look at them.

**Comment by swannman** [comment]
Resolved by merging #105
